compiler/src/iree/compiler/Dialect/Flow/Transforms/SetEncoding.cpp

```diff
@@ -21,105 +21,126 @@
 #include "mlir/Dialect/MemRef/Transforms/Transforms.h"
 #include "mlir/Dialect/Tensor/IR/Tensor.h"
 #include "mlir/Dialect/Tensor/Utils/Utils.h"
+#include "mlir/IR/BuiltinAttributeInterfaces.h"
+#include "mlir/IR/BuiltinAttributes.h"
 #include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/OpDefinition.h"
 #include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Types.h"
 #include "mlir/Transforms/GreedyPatternRewriteDriver.h"
 
 namespace mlir {
 namespace iree_compiler {
 namespace IREE {
 namespace Flow {
 
-using IREE::LinalgExt::EncodingRole;
-using IREE::LinalgExt::EncodingUser;
-
 //===---------------------------------------------------------------------===//
 // Utility functions
 //===---------------------------------------------------------------------===//
 
-// Returns the element type of `t` if it is a `ShapedType`, else return
-// `t` itself.
-static Type getElementTypeOrType(Type t) {
-  if (auto shapedType = llvm::dyn_cast<ShapedType>(t)) {
-    return shapedType.getElementType();
+/// Pads `value` enough for any actual tile sizes that could result from
+/// materialization of `encodingAttr`.
+static Value pad(OpBuilder &builder, Location loc, Value source,
```

Comment:
This is the part of the PR that I think was being discussed. Following Ben's suggestion about teaching stream to use the logical shape + tensor encoding [here](https://github.com/openxla/iree/blob/3f71a2f7f95971cc28d2219428b751050343d62f/compiler/src/iree/compiler/Dialect/Stream/Transforms/EncodeTensors.cpp#L49) will make it unnecessary to use an explicit `pad` operation. We will basically need only to add a `iree_linalg_ext.set_encoding` cause the encoding effectively includes padding at that point. So then that will obviate the need for the op added here. I am sorry I suggested that we should add an op but that was under the assumption that we had that stream needs the logical shape and physical shape to correspond exactly. But Ben is suggesting that we can use logical shape + encoding to get the physical shape. That is the more robust way forward. I see there is a lot of code here that you added to make this work, but all of that might become unnecessary with Ben's suggestion. That's my mistake (I was working off prior from discussing with Ben a few months ago)

Reply:
I think we agree 100% - but my understanding of the outcome of yesterday's discussion is that even though this isn't the final state, it is useful to merge this, as:

- While this adds some code that will be thrown away, it replaces an equivalent mass of preexisting code (particularly with the follow-up PR) and it brings the design closer to what Ben wants to replace it with. It replaces some brittle hacks (the SetEncoding pass taking a defaultPadding option with a default value 16 that must be left in its default value or else mysterious crashes happen at runtime as padded buffers overflow).
- This unblocks multiple conversations (narrow matrix multiplication performance, writing microkernels for these cases).